src/lj_opt_fold.c

```diff
@@ -1260,6 +1260,7 @@ LJFOLDF(simplify_conv_narrow)
   IROp op = (IROp)fleft->o;
   IRType t = irt_type(fins->t);
   IRRef op1 = fleft->op1, op2 = fleft->op2, mode = fins->op2;
+  mode = (IRT_INT << IRCONV_DSH) | (mode & IRCONV_SRCMASK);
```

Comment:
I think this fix is wrong. Note that you have IR:

```
left src_t ARITH  op1 op2
fins dst_t CONV   left dst_t.src_t
```

and you want to push conversion through arriving to IR

```
op1' dst_t CONV   op1 dst_t.src_t
op2' dst_t CONV   op2 dst_t.src_t 
fins dst_t ARITH  op1' op2'
```

With this fix we end up with IR like this:

```
op1' int CONV   op1 int.src_t
op2' int CONV   op2 int.src_t 
fins dst_t ARITH  op1' op2'
```

which does not seem right at all. 

From this it follows that the actual bug is in lines 1264/1265. They should read:

```
op1 = emitir(IRT(IR_CONV, t), op1, mode);
op2 = emitir(IRT(IR_CONV, t), op2, mode);
```

i.e. we need to preserve the type of the conversion we are pushing upwards with respect to its mode - not the other way arround - change the mode to satisfy INT type.

My guess about the origin of the bug would be that first this rule only handled `IRCONV_INT_*` cases (so using `IRTI` was correct). Then when it was extended to handle `IRCONV_U32_*` cases it was missed that `emitir` should be updated. `git blame` seems to corroborate this.

If you want I can make a PR fixing this.

Reply:
[Yikes, I somehow ended up editing your comment instead of replying, sorry about that!]

> My guess about the origin of the bug would be that first this rule only handled `IRCONV_INT_*` cases (so using `IRTI` was correct). Then when it was extended to handle `IRCONV_U32_*` cases it was missed that `emitir` should be updated. `git blame` seems to corroborate this.
>
> If you want I can make a PR fixing this.  

Thanks for digging up the history.  I agree that fixing up the mode to preserve sign is a better idea.  Please make a PR for this.